22_CppCompat/main.cpp

```diff
@@ -366,6 +416,358 @@ class CompatibilityTest final : public MonoDeviceApplication, public MonoAssetMa
+            static_assert(StartedValue < FinishedValue);
+            progress = m_device->createSemaphore(StartedValue);
+            {
+                IQueue* queue = m_device->getQueue(queueFamily, 0);
+
+                IQueue::SSubmitInfo submitInfos[1] = {};
+                const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufs[] = { {.cmdbuf = cmdbuf.get()} };
+                submitInfos[0].commandBuffers = cmdbufs;
+                const IQueue::SSubmitInfo::SSemaphoreInfo signals[] = { {.semaphore = progress.get(),.value = FinishedValue,.stageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+                submitInfos[0].signalSemaphores = signals;
+
+                queue->startCapture();
+                queue->submit(submitInfos);
+                queue->endCapture();
+            }
+        }
+
+        // As the name implies this function will not progress until the fence signals or repeated waiting returns an error.
+        const ISemaphore::SWaitInfo waitInfos[] = { {
+            .semaphore = progress.get(),
+            .value = FinishedValue
+        } };
+        m_device->blockForSemaphores(waitInfos);
+
+        // if the mapping is not coherent the range needs to be invalidated to pull in new data for the CPU's caches
+        const ILogicalDevice::MappedMemoryRange memoryRange(allocation.memory.get(), 0ull, allocation.memory->getAllocationSize());
+        if (!allocation.memory->getMemoryPropertyFlags().hasFlags(IDeviceMemoryAllocation::EMPF_HOST_COHERENT_BIT))
+            m_device->invalidateMappedMemoryRanges(1, &memoryRange);
+
+        assert(memoryRange.valid() && memoryRange.length >= sizeof(TestValues));
+        TestValues expectedTestValues = {
+            .intCreateVal = static_cast<emulated::emulated_float64_t::storage_t>(24.0),
+            .uintCreateVal = static_cast<emulated::emulated_float64_t::storage_t>(24u),
+            .uint64CreateVal = static_cast<emulated::emulated_float64_t::storage_t>(24ull),
+            .floatCreateVal = static_cast<emulated::emulated_float64_t::storage_t>(1.2f),
+            .doubleCreateVal = static_cast<emulated::emulated_float64_t::storage_t>(1.2),
+            .additionVal = emulated::emulated_float64_t::create(30.0),
+            .substractionVal = emulated::emulated_float64_t::create(10.0),
+            .multiplicationVal = emulated::emulated_float64_t::create(200.0),
+            .divisionVal = emulated::emulated_float64_t::create(2.0),
+            .lessOrEqualVal = false,
+            .greaterOrEqualVal = true,
+            .equalVal = false,
+            .notEqualVal = true,
+            .lessVal = false,
+            .greaterVal = true,
+            .convertionToBoolVal = true,
+            .convertionToIntVal = 20,
+            .convertionToUint32Val = 20u,
+            .convertionToUint64Val = 20ull,
+            .convertionToFloatVal = 20.0f,
+            .convertionToDoubleVal = 20.0,
+            //.convertionToHalfVal = 20;
+        };
```

Comment:
what you could do is use xoroshiro64** or somethign similar to come up with 64bit patterns denoting individual lhs and rhs values, then runs same code on CPU and GPU for a few hundred/thousand "random" doubles and see if you match

Obviously the `create` with different data types, and handling of Inf and NaN in operators needs to be tested specially, but you get the flow.

The more of the operation/input matching code you can hoist to a common HLSL/C++ header the better.